### PR TITLE
Widget Fulfillment Department

### PR DESCRIPTION
**Pull request by chuck-dbos**
Widget Fulfillment Department app

Uses Kafka to get orders from Widget Store, and presents a very simple UI for marking them as fulfilled.  Allows crashing the app to see that nothing ever gets lost.

### CODE REVIEW

File: widget-fulfillment/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# DBOS Widget Store - Fulfillment Department
+
+Visit the main [widget store](../widget-store/README.md) app for an opportunity to buy a widget!
+
+After payment is complete, your order comes here for fulfillment (if you have Kafka working, that is, so see the instructions in the Widget Store for setting it up).
```
Comment (apoliakov) [inline widget-fulfillment/README.md:5]:
```suggestion
After payment is complete, your order comes here for fulfillment (if you set up Kafka as shown at https://github.com/dbos-inc/dbos-demo-apps/tree/main/widget-store#readme).
```

File: widget-fulfillment/src/utilities.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+  @Transaction()
+  static async getUserAssignment(ctx: KnexTransactionContext, packer_name: string, expiration: Date, @ArgOptional more_time: boolean | undefined) {
+    let packers = await ctx.client<Packer>('packer').where({packer_name}).select();
+    let newAssignment = false;
+    if (!packers.length) {
+      await ctx.client<Packer>('packer').insert({packer_name, order_id: null, expiration: null});
+      packers = await ctx.client<Packer>('packer').where({packer_name}).select();
+    }
+    if (packers[0].order_id && more_time) {
```
Comment (apoliakov) [inline widget-fulfillment/src/utilities.ts:86]: We always expect to have 0 or one packer here, right? This is more of a question - is it common to use the plural form in such cases?

File: widget-store/src/operations.ts [hunk 3]
```diff
@@ -1,11 +1,30 @@
...
 export const ORDER_URL_EVENT = "order_url";
 
+const kafkaConfig: KafkaConfig = {
+  clientId: 'dbos-kafka-test',
+  brokers: [`${process.env['KAFKA_BROKER'] ?? 'localhost:9092'}`],
+  requestTimeout: 100, // FOR TESTING
+  retry: { // FOR TESTING
+    retries: 5
+  },
+  logLevel: logLevel.NOTHING, // FOR TESTING
```
Comment (apoliakov) [inline widget-store/src/operations.ts:17]: Would it be better to do like logLevel.ERROR?

File: widget-fulfillment/src/utilities.ts (hunk 2 above)
Comment (chuck-dbos) [reply widget-fulfillment/src/utilities.ts:86]: For me it's common to use the plural when it's an array... zero or more.

File: widget-store/src/operations.ts (hunk 3 above)
Comment (chuck-dbos) [reply widget-store/src/operations.ts:17]: That is a copypaste from somewhere.  I'll commit whatever level you want.

File: widget-store/src/utilities.ts [hunk 4]
```diff
@@ -6,7 +6,8 @@ type KnexTransactionContext = TransactionContext<Knex>;
 
 export enum OrderStatus {
   PENDING = 0,
-  FULFILLED = 1,
+  DISPATCHED = 1,
+  PAID = 2,
```
Comment (apoliakov) [inline widget-store/src/utilities.ts:10]: This is a new state, but it seems to briefly exist... after payment confirmation and before the kafka message is sent. So it seems to be mostly cosmetic in nature? Or did I miss something?

File: widget-store/src/operations.ts (hunk 3 above)
Comment (apoliakov) [reply widget-store/src/operations.ts:17]: Yea your guess is good as mine... I was thinking like "what would an example production app have here?"

**Review by apoliakov — APPROVED** [review]

File: widget-store/src/utilities.ts (hunk 4 above)
Comment (chuck-dbos) [reply widget-store/src/utilities.ts:10]: If the kafka message failed it'd get left that way.
In a more serious (more complex / worse for demo) you wouldn't move on until you got a message from fulfillment.

**Pull request merged**